packages/gatsby/src/schema/schema.js

```diff
@@ -1021,46 +1016,39 @@ const addImplicitConvenienceChildrenFields = ({
   const childNodesByType = groupChildNodesByType({ nodes })
 
   Object.keys(childNodesByType).forEach(typeName => {
-    const typeChildren = childNodesByType[typeName]
-    const maxChildCount = _.maxBy(
-      _.values(_.groupBy(typeChildren, c => c.parent)),
-      g => g.length
-    ).length
-
     // Adding children fields to types with the `@dontInfer` extension is deprecated
     if (shouldInfer === false) {
       const childTypeComposer = schemaComposer.getAnyTC(typeName)
       const childOfExtension = childTypeComposer.getExtension(`childOf`)
-      const many = maxChildCount > 1
 
       // Only warn when the parent-child relation has not been explicitly set with
       if (
         !childOfExtension ||
-        !childOfExtension.types.includes(parentTypeName) ||
-        !childOfExtension.many === many
+        !childOfExtension.types.includes(parentTypeName)
       ) {
-        const fieldName = many
-          ? fieldNames.convenienceChildren(typeName)
-          : fieldNames.convenienceChild(typeName)
-        const manyArg = many ? `, many: true` : ``
+        const childField = fieldNames.convenienceChild(typeName)
+        const childrenField = fieldNames.convenienceChildren(typeName)
+        const childOfTypes = (childOfExtension?.types ?? [])
+          .concat(parentTypeName)
+          .map(name => `"${name}"`)
+          .join(`,`)
+
         report.warn(
           `Deprecation warning: ` +
-            `In Gatsby v3 field \`${parentTypeName}.${fieldName}\` will not be added automatically because ` +
+            `In Gatsby v3 fields \`${parentTypeName}.${childField}\` and \`${parentTypeName}.${childrenField}\` ` +
+            `will not be added automatically because ` +
             `type \`${typeName}\` does not explicitly list type \`${parentTypeName}\` in \`childOf\` extension.\n` +
             `Add the following type definition to fix this:\n\n` +
-            `  type ${typeName} implements Node @childOf(types: ["${parentTypeName}"]${manyArg}) {\n` +
+            `  type ${typeName} implements Node @childOf(types: [${childOfTypes}]) {\n` +
             `    id: ID!\n` +
             `  }\n\n` +
             `https://www.gatsbyjs.com/docs/actions/#createTypes`
         )
       }
     }
 
-    if (maxChildCount > 1) {
-      typeComposer.addFields(createChildrenField(typeName))
-    } else {
-      typeComposer.addFields(createChildField(typeName))
-    }
+    typeComposer.addFields(createChildrenField(typeName))
+    typeComposer.addFields(createChildField(typeName))
```

Comment:
And one more entry. Other changes are mostly chores and don't have any effect on behavior.